File: client/src/types/calendar.ts
//! Calendar domain types

export interface Calendar {
  id: string;
  owner_id: string;
  name: string;
  description?: string;
  timezone: string;
  color: string;
  is_shared: boolean;
  is_public: boolean;
  created_at: string;
  updated_at: string;
  tenant_id?: string;
  workspace_id?: string;
  calendar_type?: string;
  resource_id?: string;
  is_default?: boolean;
  deleted_at?: string;
}

export interface CreateCalendar {
  name: string;
  description?: string;
  timezone?: string;
  color?: string;
  is_shared?: boolean;
}

export interface UpdateCalendar {
  name?: string;
  description?: string;
  timezone?: string;
  color?: string;
  is_shared?: boolean;
}

export interface EventAttendeeBasic {
  email?: string;
  name?: string;
}

export interface Event {
  id: string;
  calendar_id: string;
  title: string;
  description?: string;
  location?: string;
  start_time: string; // ISO 8601
  end_time: string; // ISO 8601
  /** Alternative start field from some calendar providers (Google, etc.) */
  start?: string | { dateTime?: string };
  /** Alternative end field from some calendar providers (Google, etc.) */
  end?: string | { dateTime?: string };
  rrule?: string; // RFC 5545
  rrule_exceptions?: string[];
  timezone: string;
  created_by: string;
  is_all_day: boolean;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  /** Attendees list — returned by some calendar API endpoints */
  attendees?: EventAttendeeBasic[];
  /** Arbitrary metadata stored on the event (check-in status, room info, etc.) */
  metadata?: Record<string, unknown>;
  // Unified event type fields
  event_type?: EventType;
  scope?: EventScope;
  status?: EventStatus;
  priority?: EventPriority;
  parent_event_id?: string;
  resource_id?: string;
  category_id?: string;
  leave_type?: LeaveType;
  presence_mode?: PresenceMode;
  approval_by?: string;
  approval_comment?: string;
  energy_level?: EnergyLevel;
  cron_expression?: string;
  cron_target?: string;
  assigned_to?: string;
  project_id?: string;
  tags?: string[];
}

export interface CreateEvent {
  title: string;
  description?: string;
  location?: string;
  start_time: string;
  end_time: string;
  rrule?: string;
  timezone?: string;
  is_all_day?: boolean;
  // Unified event type fields
  event_type?: string;
  scope?: string;
  status?: string;
  priority?: string;
  parent_event_id?: string;
  resource_id?: string;
  category_id?: string;
  leave_type?: string;
  presence_mode?: string;
  approval_by?: string;
  approval_comment?: string;
  energy_level?: string;
  cron_expression?: string;
  cron_target?: string;
  assigned_to?: string;
  project_id?: string;
  tags?: string[];
}

export interface UpdateEvent {
  title?: string;
  description?: string;
  location?: string;
  start_time?: string;
  end_time?: string;
  rrule?: string;
  timezone?: string;
  is_all_day?: boolean;
  // Unified event type fields
  event_type?: string;
  scope?: string;
  status?: string;
  priority?: string;
  parent_event_id?: string;
  resource_id?: string;
  category_id?: string;
  leave_type?: string;
  presence_mode?: string;
  approval_by?: string;
  approval_comment?: string;
  energy_level?: string;
  cron_expression?: string;
  cron_target?: string;
  assigned_to?: string;
  project_id?: string;
  tags?: string[];
}

export interface EventAttendee {
  id: string;
  event_id: string;
  user_id?: string;
  email?: string;
  rsvp_status: "pending" | "accepted" | "declined" | "tentative";
  response_date?: string;
  created_at: string;
  updated_at: string;
}

export interface AddEventAttendee {
  user_id?: string;
  email?: string;
}

export interface CalendarMember {
  id: string;
  calendar_id: string;
  user_id: string;
  role: "owner" | "editor" | "viewer";
  created_at: string;
  updated_at: string;
}

export type EventType =
  | "event"
  | "task"
  | "leave"
  | "shift"
  | "booking"
  | "milestone"
  | "blocker"
  | "cron";
export type EventScope = "personal" | "team" | "org";
export type EventStatus =
  | "draft"
  | "pending"
  | "approved"
  | "rejected"
  | "completed";
export type EventPriority = "low" | "medium" | "high" | "urgent";
export type LeaveType = "cp" | "rtt" | "sick" | "unpaid" | "other";
export type PresenceMode = "office" | "remote" | "absent";
export type EnergyLevel = "low" | "medium" | "high";
export type RuleType =
  | "min_onsite"
  | "mandatory_days"
  | "max_remote_same_day"
  | "min_coverage";
export type EnforcementLevel = "soft" | "hard";

export interface Category {
  id: string;
  name: string;
  color: string;
  icon?: string;
  owner_id?: string;
  org_id?: string;
  rules: Record<string, unknown>;
}

export interface PresenceRule {
  id: string;
  org_id: string;
  team_id?: string;
  rule_type: RuleType;
  rule_config: Record<string, unknown>;
  enforcement: EnforcementLevel;
  active: boolean;
}

export interface LeaveBalance {
  id: string;
  user_id: string;
  leave_type: LeaveType;
  year: number;
  total_days: number;
  used_days: number;
  pending_days: number;
}

export interface TimesheetEntry {
  id: string;
  user_id: string;
  event_id?: string;
  date: string;
  hours: number;
  category_id?: string;
  auto_generated: boolean;
  validated: boolean;
  validated_at?: string;
  exported_at?: string;
}

export interface LayerConfig {
  layer_id: string;
  enabled: boolean;
  opacity: number;
  color_override?: string;
}

export interface HeadcountPoint {
  time: string;
  role: string;
  count: number;
}

export interface PresenceViolation {
  rule_id: string;
  rule_type: RuleType;
  enforcement: EnforcementLevel;
  message: string;
}

// ============================================================================
// Project (from tenant model)
// ============================================================================

export type ProjectStatus =
  | "planning"
  | "active"
  | "on_hold"
  | "completed"
  | "archived";

export interface Project {
  id: string;
  tenant_id: string;
  workspace_id?: string;
  name: string;
  description?: string;
  color: string;
  icon?: string;
  status: ProjectStatus;
  start_date?: string;
  due_date?: string;
  calendar_id?: string;
  template_id?: string;
  owner_id?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface CreateProject {
  workspace_id?: string;
  name: string;
  description?: string;
  color?: string;
  icon?: string;
  status?: string;
  start_date?: string;
  due_date?: string;
  template_id?: string;
}

export interface UpdateProject {
  workspace_id?: string;
  name?: string;
  description?: string;
  color?: string;
  icon?: string;
  status?: string;
  start_date?: string;
  due_date?: string;
  metadata?: Record<string, unknown>;
}

export interface ProjectWithStats {
  id: string;
  tenant_id: string;
  workspace_id?: string;
  name: string;
  description?: string;
  color: string;
  status: string;
  start_date?: string;
  due_date?: string;
  owner_id?: string;
  created_at: string;
  total_tasks: number;
  completed_tasks: number;
}

// ============================================================================
// TenantTask (from tenant model)
// ============================================================================

export interface TenantTask {
  id: string;
  tenant_id: string;
  calendar_id: string;
  project_id?: string;
  parent_task_id?: string;
  event_id?: string;
  title: string;
  description?: string;
  status: string;
  priority: number;
  position: number;
  due_date?: string;
  estimated_hours?: number;
  assigned_to?: string;
  template_id?: string;
  created_by: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateTenantTask {
  calendar_id: string;
  project_id?: string;
  parent_task_id?: string;
  title: string;
  description?: string;
  priority?: number;
  position?: number;
  due_date?: string;
  estimated_hours?: number;
  assigned_to?: string;
  template_id?: string;
}

File: crates/signapps-db/src/models/calendar.rs
//! Calendar domain models (events, tasks, resources, sharing)

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ============================================================================
// Calendar Model
// ============================================================================
/// A personal or shared calendar belonging to a user.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct Calendar {
    pub id: Uuid,
    pub owner_id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub timezone: String,
    pub color: String,
    pub is_shared: bool,
    pub is_public: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub tenant_id: Option<Uuid>,
    pub workspace_id: Option<Uuid>,
    pub calendar_type: Option<String>,
    pub resource_id: Option<Uuid>,
    pub is_default: Option<bool>,
    pub deleted_at: Option<DateTime<Utc>>,
}

/// Request to create a new calendar.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CreateCalendar {
    pub name: String,
    pub description: Option<String>,
    pub timezone: Option<String>,
    pub color: Option<String>,
    pub is_shared: Option<bool>,
}

/// Request to update an existing calendar.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct UpdateCalendar {
    pub name: Option<String>,
    pub description: Option<String>,
    pub timezone: Option<String>,
    pub color: Option<String>,
    pub is_shared: Option<bool>,
}

// ============================================================================
// Calendar Member (Sharing)
// ============================================================================
/// A user who has been granted access to a shared calendar with a specific role.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CalendarMember {
    pub id: Uuid,
    pub calendar_id: Uuid,
    pub user_id: Uuid,
    pub role: String, // owner|editor|viewer
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Request to add a user to a calendar with a given role.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct AddCalendarMember {
    pub user_id: Uuid,
    pub role: String,
}

/// A calendar together with its full member list.
#[derive(Debug, Clone, Serialize)]
pub struct CalendarWithMembers {
    pub calendar: Calendar,
    pub members: Vec<CalendarMemberWithUser>,
}

/// A calendar member row joined with basic user profile data.
#[derive(Debug, Clone, FromRow, Serialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CalendarMemberWithUser {
    pub id: Uuid,
    pub user_id: Uuid,
    pub username: String,
    pub email: Option<String>,
    pub role: String,
}

// ============================================================================
// Event Model
// ============================================================================
/// A calendar event with optional recurrence rules, type metadata, and attendee support.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct Event {
    pub id: Uuid,
    pub calendar_id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub location: Option<String>,
    pub start_time: DateTime<Utc>,
    pub end_time: DateTime<Utc>,
    pub rrule: Option<String>,
    pub rrule_exceptions: Vec<Uuid>,
    pub timezone: String,
    pub created_by: Uuid,
    pub is_all_day: bool,
    pub is_deleted: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    // Unified event type fields
    pub event_type: Option<String>,
    pub scope: Option<String>,
    pub status: Option<String>,
    pub priority: Option<String>,
    pub parent_event_id: Option<Uuid>,
    pub resource_id: Option<Uuid>,
    pub category_id: Option<Uuid>,
    pub leave_type: Option<String>,
    pub presence_mode: Option<String>,
    pub approval_by: Option<Uuid>,
    pub approval_comment: Option<String>,
    pub energy_level: Option<String>,
    pub cron_expression: Option<String>,
    pub cron_target: Option<String>,
    pub assigned_to: Option<Uuid>,
    pub project_id: Option<Uuid>,
    pub tags: Option<Vec<String>>,
}

/// Request to create a new calendar event.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CreateEvent {
    pub title: String,
    pub description: Option<String>,
    pub location: Option<String>,
    pub start_time: DateTime<Utc>,
    pub end_time: DateTime<Utc>,
    pub rrule: Option<String>,
    pub timezone: Option<String>,
    pub is_all_day: Option<bool>,
    // Unified event type fields
    pub event_type: Option<String>,
    pub scope: Option<String>,
    pub status: Option<String>,
    pub priority: Option<String>,
    pub parent_event_id: Option<Uuid>,
    pub resource_id: Option<Uuid>,
    pub category_id: Option<Uuid>,
    pub leave_type: Option<String>,
    pub presence_mode: Option<String>,
    pub approval_by: Option<Uuid>,
    pub approval_comment: Option<String>,
    pub energy_level: Option<String>,
    pub cron_expression: Option<String>,
    pub cron_target: Option<String>,
    pub assigned_to: Option<Uuid>,
    pub project_id: Option<Uuid>,
    pub tags: Option<Vec<String>>,
}

/// Request to update an existing calendar event.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct UpdateEvent {
    pub title: Option<String>,
    pub description: Option<String>,
    pub location: Option<String>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub rrule: Option<String>,
    pub timezone: Option<String>,
    pub is_all_day: Option<bool>,
    // Unified event type fields
    pub event_type: Option<String>,
    pub scope: Option<String>,
    pub status: Option<String>,
    pub priority: Option<String>,
    pub parent_event_id: Option<Uuid>,
    pub resource_id: Option<Uuid>,
    pub category_id: Option<Uuid>,
    pub leave_type: Option<String>,
    pub presence_mode: Option<String>,
    pub approval_by: Option<Uuid>,
    pub approval_comment: Option<String>,
    pub energy_level: Option<String>,
    pub cron_expression: Option<String>,
    pub cron_target: Option<String>,
    pub assigned_to: Option<Uuid>,
    pub project_id: Option<Uuid>,
    pub tags: Option<Vec<String>>,
}

// ============================================================================
// Event with Attendees & Resources
// ============================================================================
/// An event bundled with its attendees, booked resources, and custom metadata.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct EventWithDetails {
    pub event: Event,
    pub attendees: Vec<EventAttendee>,
    pub resources: Vec<EventResource>,
    pub metadata: Vec<EventMetadata>,
}

// ============================================================================
// Event Attendee (RSVP)
// ============================================================================
/// An attendee (user or external email) invited to a calendar event with an RSVP status.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct EventAttendee {
    pub id: Uuid,
    pub event_id: Uuid,
    pub user_id: Option<Uuid>,
    pub email: Option<String>,
    pub rsvp_status: String, // pending|accepted|declined|tentative
    pub response_date: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Request to add an attendee to an event.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct AddEventAttendee {
    pub user_id: Option<Uuid>,
    pub email: Option<String>,
}

/// Request to update an attendee's RSVP status.
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct UpdateAttendeeRsvp {
    pub rsvp_status: String,
}

// ============================================================================
// Event Metadata
// ============================================================================
/// A key-value metadata entry attached to a calendar event.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct EventMetadata {
    pub id: Uuid,
    pub event_id: Uuid,
    pub key: String,
    pub value: Option<String>,
    pub created_at: DateTime<Utc>,
}

// ============================================================================
// Resource
// ============================================================================
/// A bookable resource such as a room, piece of equipment, or vehicle.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct Resource {
    pub id: Uuid,
    pub name: String,
    pub resource_type: String, // room|equipment|vehicle
    pub description: Option<String>,
    pub capacity: Option<i32>,
    pub location: Option<String>,
    pub is_available: bool,
    pub owner_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Request to create a new bookable resource.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateResource {
    pub name: String,
    pub resource_type: String,
    pub description: Option<String>,
    pub capacity: Option<i32>,
    pub location: Option<String>,
}

// ============================================================================
// Event Resource (booking)
// ============================================================================
/// A link between a calendar event and a booked resource.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct EventResource {
    pub id: Uuid,
    pub event_id: Uuid,
    pub resource_id: Uuid,
    pub created_at: DateTime<Utc>,
}

// ============================================================================
// Task (Hierarchical with parent_id)
// ============================================================================
/// A hierarchical task associated with a calendar, optionally assigned to a user.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct Task {
    pub id: Uuid,
    pub calendar_id: Uuid,
    pub parent_task_id: Option<Uuid>,
    pub title: String,
    pub description: Option<String>,
    pub status: String, // open|in_progress|completed|archived
    pub priority: i32,  // 0=low, 1=medium, 2=high, 3=urgent
    pub position: i32,
    pub due_date: Option<NaiveDate>,
    pub assigned_to: Option<Uuid>,
    pub created_by: Uuid,
    pub completed_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Request to create a new task.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateTask {
    pub parent_task_id: Option<Uuid>,
    pub title: String,
    pub description: Option<String>,
    pub priority: Option<i32>,
    pub position: Option<i32>,
    pub due_date: Option<NaiveDate>,
    pub assigned_to: Option<Uuid>,
}

/// Request to update an existing task.
#[derive(Debug, Clone, Deserialize)]
pub struct UpdateTask {
    pub title: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub priority: Option<i32>,
    pub position: Option<i32>,
    pub due_date: Option<NaiveDate>,
    pub assigned_to: Option<Uuid>,
}

// ============================================================================
// Task Tree (with children)
// ============================================================================
/// A task node in a recursive tree structure, containing child tasks and attachments.
#[derive(Debug, Clone, Serialize)]
pub struct TaskNode {
    pub task: Task,
    pub children: Vec<TaskNode>,
    pub attachments: Vec<TaskAttachment>,
}

// ============================================================================
// Task Attachment
// ============================================================================
/// A file attached to a task, referenced by URL.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct TaskAttachment {
    pub id: Uuid,
    pub task_id: Uuid,
    pub file_url: String,
    pub file_name: Option<String>,
    pub file_size_bytes: Option<i32>,
    pub created_at: DateTime<Utc>,
}

// ============================================================================
// Reminder
// ============================================================================
/// A scheduled reminder for an event or task, delivered via notification, email, or SMS.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct Reminder {
    pub id: Uuid,
    pub event_id: Option<Uuid>,
    pub task_id: Option<Uuid>,
    pub user_id: Uuid,
    pub reminder_type: String, // notification|email|sms
    pub minutes_before: i32,
    pub is_sent: bool,
    pub sent_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

/// Request to create a reminder for an event or task.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateReminder {
    pub event_id: Option<Uuid>,
    pub task_id: Option<Uuid>,
    pub user_id: Uuid,
    pub reminder_type: Option<String>,
    pub minutes_before: Option<i32>,
}

// ============================================================================
// Activity Log (Audit Trail)
// ============================================================================
/// An audit trail entry recording a change to a calendar entity.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ActivityLog {
    pub id: Uuid,
    pub calendar_id: Uuid,
    pub entity_type: String, // event|task|calendar|resource
    pub entity_id: Uuid,
    pub action: String, // created|updated|deleted|shared
    pub user_id: Option<Uuid>,
    pub changes: Option<serde_json::Value>,
    pub created_at: DateTime<Utc>,
}

// ============================================================================
// FloorPlan Models
// ============================================================================
/// A floor plan layout used for visualising and booking physical resources.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct FloorPlan {
    pub id: Uuid,
    pub name: String,
    pub floor: String,
    pub width: f64,
    pub height: f64,
    pub resources: serde_json::Value,
    pub svg_content: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Request to create a new floor plan.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateFloorPlan {
    pub name: String,
    pub floor: String,
    pub width: f64,
    pub height: f64,
    pub resources: serde_json::Value,
    #[serde(rename = "svgContent")]
    pub svg_content: Option<String>,
}

/// Request to update an existing floor plan.
#[derive(Debug, Clone, Deserialize)]
pub struct UpdateFloorPlan {
    pub name: Option<String>,
    pub floor: Option<String>,
    pub width: Option<f64>,
    pub height: Option<f64>,
    pub resources: Option<serde_json::Value>,
    #[serde(rename = "svgContent")]
    pub svg_content: Option<String>,
}

// ============================================================================
// Category
// ============================================================================
/// A color-coded category used to classify calendar events and tasks.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct Category {
    pub id: Uuid,
    pub name: String,
    pub color: String,
    pub icon: Option<String>,
    pub owner_id: Option<Uuid>,
    pub org_id: Option<Uuid>,
    pub rules: Option<serde_json::Value>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

/// Request to create a new event/task category.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateCategory {
    pub name: String,
    pub color: Option<String>,
    pub icon: Option<String>,
    pub org_id: Option<Uuid>,
    pub rules: Option<serde_json::Value>,
}

// ============================================================================
// PresenceRule
// ============================================================================
/// An organisational rule governing on-site vs. remote presence requirements for a team.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct PresenceRule {
    pub id: Uuid,
    pub org_id: Uuid,
    pub team_id: Option<Uuid>,
    pub rule_type: String,
    pub rule_config: serde_json::Value,
    pub enforcement: Option<String>,
    pub active: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

/// Request to create a new presence rule for an organisation or team.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreatePresenceRule {
    pub org_id: Uuid,
    pub team_id: Option<Uuid>,
    pub rule_type: String,
    pub rule_config: serde_json::Value,
    pub enforcement: Option<String>,
}

// ============================================================================
// LeaveBalance
// ============================================================================
/// A user's annual leave balance tracking total, used, and pending days per leave type.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct LeaveBalance {
    pub id: Uuid,
    pub user_id: Uuid,
    pub leave_type: String,
    pub year: i32,
    pub total_days: f64,
    pub used_days: f64,
    pub pending_days: f64,
    pub updated_at: Option<DateTime<Utc>>,
}

// ============================================================================
// TimesheetEntry
// ============================================================================
/// A timesheet entry recording hours worked by a user on a given date.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct TimesheetEntry {
    pub id: Uuid,
    pub user_id: Uuid,
    pub event_id: Option<Uuid>,
    pub date: NaiveDate,
    pub hours: f64,
    pub category_id: Option<Uuid>,
    pub auto_generated: Option<bool>,
    pub validated: Option<bool>,
    pub validated_at: Option<DateTime<Utc>>,
    pub exported_at: Option<DateTime<Utc>>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

// ============================================================================
// ApprovalWorkflow
// ============================================================================
/// An approval workflow triggered by configurable calendar events (e.g. leave requests).
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ApprovalWorkflow {
    pub id: Uuid,
    pub org_id: Uuid,
    pub trigger_type: String,
    pub trigger_config: serde_json::Value,
    pub approvers: serde_json::Value,
    pub active: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

/// Request to create a new approval workflow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateApprovalWorkflow {
    pub org_id: Uuid,
    pub trigger_type: String,
    pub trigger_config: serde_json::Value,
    pub approvers: serde_json::Value,
}
